surround/assembler.py

```diff
@@ -0,0 +1,184 @@
+# assembler.py
+#
+# Manages a set of stages and the data that is passed between them.
+import sys
+import os
+import logging
+
+from abc import ABC
+from datetime import datetime
+
+from .config import Config
+from .loader import Loader
+from .surround import SurroundData
+from .visualiser import Visualiser
+from .stage import Validator, Filter, Estimator
+
+LOGGER = logging.getLogger(__name__)
+
+class Assembler(ABC):
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self, assembler_name="", surround_data=None, estimator=None, config=None):
+        self.assembler_name = assembler_name
+        self.config = config
+        self.surround_data = surround_data
+        self.estimator = estimator
+        self.loader = None
+        self.validator = None
+        self.pre_filters = None
+        self.post_filters = None
+        self.visualiser = None
+
+    def init_assembler(self):
+        try:
+            if self.validator:
+                self.validator.init_stage(self.config)
+
+            if self.pre_filters:
+                for pre_filter in self.pre_filters:
+                    pre_filter.init_stage(self.config)
+
+            self.estimator.init_stage(self.config)
+
+            if self.post_filters:
+                for post_filter in self.post_filters:
+                    post_filter.init_stage(self.config)
+        except Exception:
+            LOGGER.exception("Failed initiating Assembler")
+
+    def run(self):
+        LOGGER.info("Starting '%s'", self.assembler_name)
+        try:
+            if self.loader:     # Initiate loader (if exists)
+                self.loader.load(self.surround_data, self.config)
+                self.validator.validate(self.surround_data, self.config)
+                self.loader.execute(self.surround_data, self.config, self.__run_pipeline)
+            else:
+                self.__run_pipeline()
+        except Exception:
+            LOGGER.exception("Failed running Assembler")
+
+    def __run_pipeline(self):
+        if self.pre_filters:
+            self.execute_filters(self.pre_filters, self.surround_data)
+
+        self.execute_main(self.surround_data)
+
+        if self.post_filters:
+            self.execute_filters(self.post_filters, self.surround_data)
+
+    def execute_filters(self, filters, surround_data):
```

Comment:
Should this and the `excute_filter` function be "public"?

Reply:
No, they are not
I'll make them private